### DISCUSSION TITLE
revalidate related field when using Joi

### DISCUSSION THREAD

**Discussion by mlarcher** [mantinedev/mantine#2630] [Q&A]
We have a form that needs to set "required" (actually `.min(1)`) on a field only when another one contains a given value.

We use JoiResolver to set up the validation rules, and it works well at first, but the related field doesn't update when the main field is altered. 

To be more precise, we want fieldB of type array to require at least one item if fieldA of type array contains one of a specific set of values. We use the following structure:

```js
validate: joiResolver(
  Joi.object({
    fieldA: Joi.array().items(Joi.string()).min(1),
    fieldB: Joi.array().items(Joi.string()).when('fieldA', {
          is: Joi.array().items(
            Joi.string(),
            Joi.string()
              .valid(...listOfFieldATriggers)
              .required(),
          ),
          then: Joi.array().items(Joi.string()).min(1),
          otherwise: Joi.array().items(Joi.string()),
        }),
```

When we submit the form, the fieldB which is empty is in error or not according to the fieldA content, but when we then change the fieldA value to something that doesn't trigger the `min(1)` rule on the fieldB, fieldB stays in error.

How come the fieldB rule doesn't update as expected ? Is there anything we can do to make it happen ?

**Comment by rtivital** [comment]
joiResolver works with errors returned by your schema, first thing to do is to make sure that schema work correctly on its own. If the default joi resolver does now satisfy your requirements, you can easily create your own resolver, see example – https://github.com/mantinedev/mantine/blob/master/src/mantine-form/src/resolvers/joi-resolver/joi-resolver.ts

**Comment by mlarcher** [comment]
Thanks for the info, it helped me dig further. 

What I can now tell is that my schema is always returning the correct error. But the issue is that form.errors does not reflect that error state until I click on the submit button again. 

Is there something to activate to get live errors ? I couldn't find anything like it in the documentation.

**Reply by rtivital** [reply]
https://mantine.dev/form/validation/#validate-fields-on-change

**Reply by mlarcher** [reply]
I did use it, the validations occurs (the function returned by the joiResolver is called and returns the correct errors) but the form.errors is not updated until I click on the submit button again.

**Reply by rtivital** [reply]
If you send a sandbox with **minimal** reproduction, I can have a look

**Reply by mlarcher** [reply]
That's nice of you :)
here it is : https://codesandbox.io/s/immutable-https-nvg5nq?file=/src/App.tsx

**Reply by rtivital** [reply]
Your sandbox is empty

**Reply by mlarcher** [reply]
it's fixed now

**Reply by rtivital** [reply]
I do not know where the issue is, it works exactly as described in your schema – https://codesandbox.io/s/objective-matan-npj0he?file=/src/App.tsx


https://user-images.githubusercontent.com/10353856/194697458-097e7fef-9057-44f1-a873-69b55a7e3940.mov

**Reply by mlarcher** [reply]
What i would expect is that after field B is in error, if I change field A to not trigger the .min() condition, the error on field B would disappear
Schema does not give error in that case, but form.errors doesn't update to reflect it

**Reply by rtivital** [reply]
Well, it seems pretty complicated and may require lots of debugging. In this case, I'd just use function instead of schema.

**Reply by mlarcher** [reply]
I don't understand what difference that would make. Isn't joiResolver such a function? It uses joi but could not do so... And it handles the errors correctly and outputs errors as expected. 

From what I see, it seems to me that further down the line, mantine form does not update its error object, or does not re render, which is why the displayed error it's not correctly updated (console.logs on the sandbox should highlight the offset between joiResolver return and form.errors values)

**Reply by mlarcher** [reply]
Or is there something I'm missing on my side that we should be doing to let form.errors update according to the result of the validator ?
